### DISCUSSION TITLE
How to improve AI task pickup speed in AIO

### DISCUSSION THREAD

**Discussion by szaimen** [nextcloud/all-in-one#5430] [Wiki]
Most AI tasks will be run as part of the background job system in Nextcloud which only runs jobs every 5 minutes by default. To pick up scheduled jobs faster you can set up background job workers that process AI tasks as soon as they are scheduled. If the PHP code or the Nextcloud settings values are changed while a worker is running, those changes won’t be effective inside the runner. For that reason, the worker needs to be restarted regularly. It is done with a timeout of N seconds which means any changes to the settings or the code will be picked up after N seconds (worst case scenario). This timeout does not, in any way, affect the processing or the timeout of the AI tasks.

Use these instructions https://docs.nextcloud.com/server/latest/admin_manual/ai/overview.html#systemd-service with two adjustments:

1. In step 1, add these three lines to the `[Unit]` section of the suggested service file:

```shell
# Ensures the service waits until Docker is fully functional
After=docker.service
Requires=docker.service
```

2. In step 2, use the following script:

```bash
#!/bin/bash
echo "Starting Nextcloud AI Worker $1"

# Function to attempt starting the worker
start_worker() {
    echo "Attempting to start Nextcloud AI Worker $1..."
    
    # Run the worker inside the container (60 seconds for task execution)
    docker exec --user www-data nextcloud-aio-nextcloud php occ background-job:worker -t 60 'OC\TaskProcessing\SynchronousBackgroundJob'
    
    # Capture the exit status of the docker exec command
    exit_code=$?

    # Check if the worker ran successfully
    if [ $exit_code -eq 0 ]; then
        echo "Worker completed successfully."
        return 0  # Success
    else
        echo "Worker failed to start or complete with exit code $exit_code."
        return 1  # Failure
    fi
}

# Loop to start the worker and retry if it fails
while true; do
    # Check if the container is running
    if [ "$(docker inspect -f '{{.State.Running}}' nextcloud-aio-nextcloud 2>/dev/null)" != "true" ]; then
        echo "Nextcloud container is not running. Waiting 60 seconds before retrying..."
        sleep 60
        continue  # Skip to the next iteration
    fi

    # Start the worker
    if ! start_worker; then
        # If failed, wait 30 seconds (or adjust as needed) to avoid rapid retries
        echo "Worker execution failed. Waiting 30 seconds before retrying..."
        sleep 30
    fi
done
```

**Comment by plainskill** [comment]
what container to run this in?

**Reply by szaimen** [reply]
Please read the instructions, especially the linked docs.

**Reply by derStephan** [reply]
I read the instructions but don't really understand what to do for AIO. 

After running 

``occ background-job:worker -v -t 60 "OC\TaskProcessing\SynchronousBackgroundJob"`` 

it really speeds up AI-things until the next reboot of the host server. 

The I tried adding 

``@reboot set -e; while true; do docker exec -it nextcloud-aio-nextcloud sudo -E -u www-data php occ background-job:worker -v -t 60 "OC\TaskProcessing\SynchronousBackgroundJob"; done``

But that did not work.  Can you explain how to set that option persistently? I can send a PR for the official documentation afterwards.

**Reply by szaimen** [reply]
You need to create a systemd service as mentioned in the docs...

**Reply by derStephan** [reply]
Ok, so I was mislead by the sentence 
>For Nextcloud-AIO you should use this command on the host server.

**Reply by derStephan** [reply]
Ok, here is what I have done for AIO **on the physical host**:

```
sudo bash
nano /root/taskprocessing.sh
```
put the content of @szaimen in and save the file, then do 
```
chmod +x /root/taskprocessing.sh
crontab -e
```
append the following and save

``
@reboot /root/taskprocessing.sh
``

This works for me on Ubuntu 24.04 without adding systemd service

**Reply by WolkeOsch** [reply]
I don't use docker so I changed...

` docker exec --user www-data nextcloud-aio-nextcloud php occ background-job:worker -t 60 'OC\TaskProcessing\SynchronousBackgroundJob'`

to

`sudo -u www-data php8.3 /path to/nextcloud/occ background-job:worker -t 60 'OC\TaskProcessing\SynchronousBackgroundJob'`

and deleted

```
    if ! docker ps --format "{{.Names}}" | grep -q "^nextcloud-aio-nextcloud$"; then
        echo "Nextcloud container is not running. Waiting 60 seconds before next attempt."
        sleep 60  # If the container isn't running, wait for 60 seconds
        continue  # Skip to the next iteration
    fi
```

then I do 
```
chmod +x /path to/taskprocessing.sh
crontab -e
@reboot /path to/taskprocessing.sh
```
This works for me on Ubuntu 24.04 without adding systemd service

**Comment by plainskill** [comment]
that's actually a good solution though. Because nextcloud aio does not have systemd(at least when i tried to use it) that could actually work, unlike the systemd way of doing it

Sent with [Proton Mail](https://proton.me/mail/home) secure email.

**Reply by derStephan** [reply]
AIO does not have to have systemd. This is to be done on the host.

**Comment by plainskill** [comment]
ohhhhhh. well that makes much more sense now.

Sent with [Proton Mail](https://proton.me/mail/home) secure email.

**Comment by szaimen** [comment]
Hi, with AIO v13 we will bundle a taskprocessing service by default. See https://github.com/nextcloud/all-in-one/pull/7981

**Reply by derStephan** [reply]
Wow. Thank you. Do we have to remove the service that we installed before?

**Reply by szaimen** [reply]
Yes you can remove it after the upgrade

**Reply by szaimen** [reply]
This is now released with v13.0.0 Beta. Testing and feedback is welcome! See https://github.com/nextcloud/all-in-one#how-to-switch-the-channel

**Reply by Fuseteam** [reply]
is there anything we have to do t make use of it after we upgrade?

**Reply by szaimen** [reply]
No it should generally work directly after the update

**Reply by DrPain911** [reply]
When will v13 be available for the latest channel?

**Reply by szaimen** [reply]
probably tomorrow

**Reply by Fuseteam** [reply]
hmmm maybe i'll wait for the upgrade. i tried the above systemd services but i still get 417

granted my set up is behind a proxy, and i only proxied 8780 :thinking:

**Reply by derStephan** [reply]
> No it should generally work directly after the update

It does not work for me. I removed the service after upgrade to AIO 13.0.4 stable and did a reboot. Now it seems that pick-up speed is back to about 5 minutes. 

When I switch the service back on, the pick up speed is as before.

**Reply by javimurcia** [reply]
The `taskprocessing` service appears to work only for the first 5 minutes after Nextcloud starts. Once the worker reaches its timeout, it exits and is not restarted, so AI tasks fall back to the normal background-job scheduling.

You can reproduce it by restarting Nextcloud and running an AI task within the first 5 minutes; it should complete quickly. Then wait until the 5-minute mark has passed and try again. At that point, the task may take up to 5 minutes to be picked up.

Until https://github.com/nextcloud/all-in-one/pull/8558 is merged and released, there is a quick workaround:

* Edit `/supervisord.conf` inside the `nextcloud-aio-nextcloud` container.
* Add `autorestart=true` to the `[program:taskprocessing-worker]` section.
* Restart the container with:

```bash
docker restart nextcloud-aio-nextcloud
```

After that, Supervisor will restart the worker whenever its 300-second timeout expires.

Note that this workaround will not survive a container recreation.

**Comment by DrPain911** [comment]
Will I have to do anything? For me the command 

set -e; while true; do docker exec -it nextcloud-aio-nextcloud sudo -E -u www-data php occ background-job:worker -v -t 60 "OC\TaskProcessing\SynchronousBackgroundJob"; done

does work to make AI reply instantaneous, but it is not preserved after reboot. Will I have to do anything besides installing AIO v13 and when is it projected to come out in the latest channel?

**Comment by neumann727** [comment]
Would it be possible to expose the timeout as docker env variable:
_command=php /var/www/html/occ taskprocessing:worker --timeout 300_
cause if it stays like this, we are probably back to 5 minutes and are likely to need some work arrounds again

**Reply by szaimen** [reply]
@marcelklehr should we rather decrease the timeout to 60 or something alike?

**Reply by marcelklehr** [reply]
> we are probably back to 5 minutes and are likely to need some work arrounds again

300s timeout means that the worker will *restart* after 5mins, it should still pick up jobs one after the other in that time. We do have a bug in that worker at the moment, that we're working to resolve, in that there's a race condition that can cause two (or more) workers to pick up the same jobs.

**Reply by Fuseteam** [reply]
i wonder if that related to what i'm seeing, seems the context backup is taking an awfull long time to index. Tis about 30k files and in a week we got to 12k........ unless this is related to GPU? as far as i can tell the cpu of the machine i'm running the context backend on is not currently stressed

**Reply by marcelklehr** [reply]
context chat uses the GPU if possible, yes. But it doesn't use the taskprocessing workers for indexing.

**Reply by Fuseteam** [reply]
oh so it's indexing as often then

**Reply by Fuseteam** [reply]
i wonder what the schedule of it is

**Reply by marcelklehr** [reply]
Please open an issue over at nextcloud/context_chat if you have specific questions or issues with it.

**Reply by Fuseteam** [reply]
will do once i have a better understanding of it

**Reply by szaimen** [reply]
Hi, apparently the service was not automatically restarted once exited. This was improved in https://github.com/nextcloud/all-in-one/pull/8558 and is now released with v13.6.0 Beta. Testing and feedback is welcome! See https://github.com/nextcloud/all-in-one#how-to-switch-the-channel
